taichi/analysis/same_statements.cpp

```diff
@@ -191,22 +203,37 @@ class IRNodeComparator : public IRVisitor {
     }
   }
 
-  static bool run(IRNode *root1, IRNode *root2) {
-    IRNodeComparator comparator(root2);
+  static bool run(IRNode *root1,
+                  IRNode *root2,
+                  std::optional<std::unordered_map<int, int>> id_map) {
+    IRNodeComparator comparator(root2, std::move(id_map));
     root1->accept(&comparator);
     return comparator.same;
   }
 };
 
 namespace irpass::analysis {
-bool same_statements(IRNode *root1, IRNode *root2) {
+bool same_statements(IRNode *root1,
+                     IRNode *root2,
+                     std::optional<std::unordered_map<int, int>> id_map) {
+  // id_map is an id map from root1 to root2.
+  //
+  // For example, same_statements($3, $6) is true
+  // iff id_map[1] == 4 && id_map[2] == 5:
+  // <i32> $3 = add $1 $2
+  // <i32> $6 = add $4 $5
+  //
+  // If capture_ids is std::nullopt by default, an identity mapping will
+  // be used. This is correct when root1 and root2 share the same IR root.
   if (root1 == root2)
     return true;
   if (!root1 || !root2)
     return false;
```

Comment:
Maybe we should let the function to fail loudly (`TI_ERROR`) in this case? The same in `same_value`.

Reply:
I think it's fine to report `nullptr` is "not same" with any `Stmt` not being a nullptr. Not sure if we used this feature anywhere in the codebase...

Reply:
This part actually utilizes the feature that `same_value(Stmt *, nullptr) == false`:
https://github.com/taichi-dev/taichi/blob/3c2e5e9b2e68362739a67407c5bf4a2eb4781a7c/taichi/ir/control_flow_graph.cpp#L165-L179